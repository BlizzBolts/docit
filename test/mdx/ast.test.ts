import { describe, it, expect } from "vitest";
import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkMdx from "remark-mdx";
import stringify from "remark-stringify";
import remarkGfm from "remark-gfm";
import remarkRehype from "remark-rehype";
import rehypeStringify from "rehype-stringify";

describe("mdxToTree", () => {
  it("markdownToAst", async (done) => {
    const ast = unified().use(remarkParse).parse("# Hi\n\n*Hello*, world!");
    expect(ast).matchSnapshot();
    done();
  });

  it("mdxToAst", async (done) => {
    const ast = unified()
      .use(remarkParse)
      .use(remarkMdx)
      .parse('import a from "b"\n\na <b /> c {1 + 1} d');
    expect(ast).matchSnapshot();
    done();
  });

  it("astToMdx", async (done) => {
    const complexString = `
      ## hello

      ## Table

      | a | b  |  c |  d  |
      | - | :- | -: | :-: |

      import { Button } from 'button'

      <Button con={1}/>
    `;
    const ast = unified().use(remarkParse).use(remarkMdx).parse(complexString);

    const result = unified()
      .use(remarkParse)
      .use(remarkMdx)
      .use(stringify)
      .use(remarkGfm)
      .stringify(ast);
    console.log(result);
    done();
  });
  it("mdxToAstWithGFM", async (done) => {
    const sample = `
    
    # GFM

    ## Autolink literals

    www.example.com, https://example.com, and contact@example.com.

    ## Footnote

    A note[^1]

    [^1]: Big note.

    ## Strikethrough

    ~one~ or ~~two~~ tildes.

    ## Table

    | a | b  |  c |  d  |
    | - | :- | -: | :-: |

    ## Tasklist

    * [ ] to do
    * [x] done
    `;

    const file = await unified()
      .use(remarkParse)
      .use(remarkMdx)
      .use(remarkGfm)
      .use(remarkRehype)
      .use(rehypeStringify)
      .process(sample);
    expect(String(file)).matchSnapshot();
    done();
  });
});
