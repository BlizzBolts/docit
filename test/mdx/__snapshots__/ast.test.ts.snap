// Vitest Snapshot v1

exports[`mdxToTree > markdownToAst 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
              "offset": 4,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hi",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 5,
          "line": 1,
          "offset": 4,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 7,
                  "line": 3,
                  "offset": 12,
                },
                "start": {
                  "column": 2,
                  "line": 3,
                  "offset": 7,
                },
              },
              "type": "text",
              "value": "Hello",
            },
          ],
          "position": {
            "end": {
              "column": 8,
              "line": 3,
              "offset": 13,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 6,
            },
          },
          "type": "emphasis",
        },
        {
          "position": {
            "end": {
              "column": 16,
              "line": 3,
              "offset": 21,
            },
            "start": {
              "column": 8,
              "line": 3,
              "offset": 13,
            },
          },
          "type": "text",
          "value": ", world!",
        },
      ],
      "position": {
        "end": {
          "column": 16,
          "line": 3,
          "offset": 21,
        },
        "start": {
          "column": 1,
          "line": 3,
          "offset": 6,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 16,
      "line": 3,
      "offset": 21,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mdxToTree > mdxToAst 1`] = `
{
  "children": [
    {
      "data": {
        "estree": Node {
          "body": [
            Node {
              "end": 17,
              "loc": {
                "end": {
                  "column": 17,
                  "line": 1,
                  "offset": 17,
                },
                "start": {
                  "column": 0,
                  "line": 1,
                  "offset": 0,
                },
              },
              "range": [
                0,
                17,
              ],
              "source": Node {
                "end": 17,
                "loc": {
                  "end": {
                    "column": 17,
                    "line": 1,
                    "offset": 17,
                  },
                  "start": {
                    "column": 14,
                    "line": 1,
                    "offset": 14,
                  },
                },
                "range": [
                  14,
                  17,
                ],
                "raw": "\\"b\\"",
                "start": 14,
                "type": "Literal",
                "value": "b",
              },
              "specifiers": [
                Node {
                  "end": 8,
                  "loc": {
                    "end": {
                      "column": 8,
                      "line": 1,
                      "offset": 8,
                    },
                    "start": {
                      "column": 7,
                      "line": 1,
                      "offset": 7,
                    },
                  },
                  "local": Node {
                    "end": 8,
                    "loc": {
                      "end": {
                        "column": 8,
                        "line": 1,
                        "offset": 8,
                      },
                      "start": {
                        "column": 7,
                        "line": 1,
                        "offset": 7,
                      },
                    },
                    "name": "a",
                    "range": [
                      7,
                      8,
                    ],
                    "start": 7,
                    "type": "Identifier",
                  },
                  "range": [
                    7,
                    8,
                  ],
                  "start": 7,
                  "type": "ImportDefaultSpecifier",
                },
              ],
              "start": 0,
              "type": "ImportDeclaration",
            },
          ],
          "comments": [],
          "end": 17,
          "loc": {
            "end": {
              "column": 17,
              "line": 1,
              "offset": 17,
            },
            "start": {
              "column": 0,
              "line": 1,
              "offset": 0,
            },
          },
          "range": [
            0,
            17,
          ],
          "sourceType": "module",
          "start": 0,
          "type": "Program",
        },
      },
      "position": {
        "end": {
          "column": 18,
          "line": 1,
          "offset": 17,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "mdxjsEsm",
      "value": "import a from \\"b\\"",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 3,
              "line": 3,
              "offset": 21,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 19,
            },
          },
          "type": "text",
          "value": "a ",
        },
        {
          "attributes": [],
          "children": [],
          "name": "b",
          "position": {
            "end": {
              "column": 8,
              "line": 3,
              "offset": 26,
            },
            "start": {
              "column": 3,
              "line": 3,
              "offset": 21,
            },
          },
          "type": "mdxJsxTextElement",
        },
        {
          "position": {
            "end": {
              "column": 11,
              "line": 3,
              "offset": 29,
            },
            "start": {
              "column": 8,
              "line": 3,
              "offset": 26,
            },
          },
          "type": "text",
          "value": " c ",
        },
        {
          "data": {
            "estree": {
              "body": [
                {
                  "end": 35,
                  "expression": Node {
                    "end": 35,
                    "left": Node {
                      "end": 31,
                      "loc": {
                        "end": {
                          "column": 12,
                          "line": 3,
                          "offset": 31,
                        },
                        "start": {
                          "column": 11,
                          "line": 3,
                          "offset": 30,
                        },
                      },
                      "range": [
                        30,
                        31,
                      ],
                      "raw": "1",
                      "start": 30,
                      "type": "Literal",
                      "value": 1,
                    },
                    "loc": {
                      "end": {
                        "column": 16,
                        "line": 3,
                        "offset": 35,
                      },
                      "start": {
                        "column": 11,
                        "line": 3,
                        "offset": 30,
                      },
                    },
                    "operator": "+",
                    "range": [
                      30,
                      35,
                    ],
                    "right": Node {
                      "end": 35,
                      "loc": {
                        "end": {
                          "column": 16,
                          "line": 3,
                          "offset": 35,
                        },
                        "start": {
                          "column": 15,
                          "line": 3,
                          "offset": 34,
                        },
                      },
                      "range": [
                        34,
                        35,
                      ],
                      "raw": "1",
                      "start": 34,
                      "type": "Literal",
                      "value": 1,
                    },
                    "start": 30,
                    "type": "BinaryExpression",
                  },
                  "loc": {
                    "end": {
                      "column": 16,
                      "line": 3,
                      "offset": 35,
                    },
                    "start": {
                      "column": 11,
                      "line": 3,
                      "offset": 30,
                    },
                  },
                  "range": [
                    30,
                    35,
                  ],
                  "start": 30,
                  "type": "ExpressionStatement",
                },
              ],
              "comments": [],
              "end": 35,
              "loc": {
                "end": {
                  "column": 16,
                  "line": 3,
                  "offset": 35,
                },
                "start": {
                  "column": 11,
                  "line": 3,
                  "offset": 30,
                },
              },
              "range": [
                30,
                35,
              ],
              "sourceType": "module",
              "start": 30,
              "type": "Program",
            },
          },
          "position": {
            "end": {
              "column": 18,
              "line": 3,
              "offset": 36,
            },
            "start": {
              "column": 11,
              "line": 3,
              "offset": 29,
            },
          },
          "type": "mdxTextExpression",
          "value": "1 + 1",
        },
        {
          "position": {
            "end": {
              "column": 20,
              "line": 3,
              "offset": 38,
            },
            "start": {
              "column": 18,
              "line": 3,
              "offset": 36,
            },
          },
          "type": "text",
          "value": " d",
        },
      ],
      "position": {
        "end": {
          "column": 20,
          "line": 3,
          "offset": 38,
        },
        "start": {
          "column": 1,
          "line": 3,
          "offset": 19,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 20,
      "line": 3,
      "offset": 38,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`mdxToTree > mdxToAstWithGFM 1`] = `
"<h1>GFM</h1>
<h2>Autolink literals</h2>
<p><a href=\\"http://www.example.com\\">www.example.com</a>, <a href=\\"https://example.com\\">https://example.com</a>, and <a href=\\"mailto:contact@example.com\\">contact@example.com</a>.</p>
<h2>Footnote</h2>
<p>A note<sup><a href=\\"#user-content-fn-1\\" id=\\"user-content-fnref-1\\" data-footnote-ref aria-describedby=\\"footnote-label\\">1</a></sup></p>
<section data-footnotes class=\\"footnotes\\"><h2 class=\\"sr-only\\" id=\\"footnote-label\\">Footnotes</h2>
<ol>
<li id=\\"user-content-fn-1\\">
<p>Big note.</p>
<h2>Strikethrough</h2>
<p><del>one</del> or <del>two</del> tildes.</p>
<h2>Table</h2>
<table>
<thead>
<tr>
<th>a</th>
<th align=\\"left\\">b</th>
<th align=\\"right\\">c</th>
<th align=\\"center\\">d</th>
</tr>
</thead>
</table>
<h2>Tasklist</h2>
<ul class=\\"contains-task-list\\">
<li class=\\"task-list-item\\"><input type=\\"checkbox\\" disabled> to do</li>
<li class=\\"task-list-item\\"><input type=\\"checkbox\\" checked disabled> done</li>
</ul>
<a href=\\"#user-content-fnref-1\\" data-footnote-backref class=\\"data-footnote-backref\\" aria-label=\\"Back to content\\">↩</a>
</li>
</ol>
</section>"
`;
